# COLLABORATIVE FILTERING - USER BASED
## BY DWI SMARADAHANA INDRALOKA
***

## Import library

In [1]:
import pandas as pd

## Upload Dataset

In [2]:
data = pd.read_csv("Film.csv")
data.head()

Timestamp         Nama Anda  Ada Apa dengan Cinta 2  \
0  2019/09/17 10:18:21 AM GMT+7             Hania                     3.0   
1  2019/09/17 10:18:37 AM GMT+7  Topik Zulkarnain                     NaN   
2  2019/09/17 10:18:39 AM GMT+7    AhokTemanFirli                     NaN   
3  2019/09/17 10:18:42 AM GMT+7          franadek                     4.0   
4  2019/09/17 10:19:01 AM GMT+7          OM INDRA                     3.0   

   Gundala  Dilan 1991  Bumi Manusia   Dua Garis Biru  Avengers: End Game  \
0      5.0         4.0            4.0             4.0                 NaN   
1      NaN         NaN            NaN             NaN                 5.0   
2      NaN         NaN            NaN             NaN                 3.0   
3      4.0         4.0            5.0             3.0                 5.0   
4      NaN         2.0            NaN             5.0                 5.0   

   The Lion King  Aladdin  Spiderman: Far From Home  Captain Marvel  
0            NaN      NaN                       NaN             NaN  
1            5.0      NaN                       4.0             2.0  
2            NaN      NaN                       NaN             4.0  
3            4.0      5.0                       4.0             4.0  
4            NaN      1.0                       5.0             5.0

## Drop Unused Column

In [3]:
data.drop("Timestamp", axis = 1, inplace = True)

## Change Column "Nama Anda" into Index

In [4]:
col = data["Nama Anda"]

In [5]:
data.set_index("Nama Anda", inplace = True)
data.head()

Ada Apa dengan Cinta 2  Gundala  Dilan 1991  Bumi Manusia   \
Nama Anda                                                                      
Hania                                3.0      5.0         4.0            4.0   
Topik Zulkarnain                     NaN      NaN         NaN            NaN   
AhokTemanFirli                       NaN      NaN         NaN            NaN   
franadek                             4.0      4.0         4.0            5.0   
OM INDRA                             3.0      NaN         2.0            NaN   

                  Dua Garis Biru  Avengers: End Game  The Lion King  Aladdin  \
Nama Anda                                                                      
Hania                        4.0                 NaN            NaN      NaN   
Topik Zulkarnain             NaN                 5.0            5.0      NaN   
AhokTemanFirli               NaN                 3.0            NaN      NaN   
franadek                     3.0                 5.0            4.0      5.0   
OM INDRA                     5.0                 5.0            NaN      1.0   

                  Spiderman: Far From Home  Captain Marvel  
Nama Anda                                                   
Hania                                  NaN             NaN  
Topik Zulkarnain                       4.0             2.0  
AhokTemanFirli                         NaN             4.0  
franadek                               4.0             4.0  
OM INDRA                               5.0             5.0

## Fill NaN Value with 0

In [6]:
data.isna().sum()

Ada Apa dengan Cinta 2       7
Gundala                     11
Dilan 1991                  10
Bumi Manusia                18
Dua Garis Biru              15
Avengers: End Game           5
The Lion King               13
Aladdin                     10
Spiderman: Far From Home     8
Captain Marvel               6
dtype: int64

In [7]:
data = data.fillna(0)
data_ori = data
data.head()

Ada Apa dengan Cinta 2  Gundala  Dilan 1991  Bumi Manusia   \
Nama Anda                                                                      
Hania                                3.0      5.0         4.0            4.0   
Topik Zulkarnain                     0.0      0.0         0.0            0.0   
AhokTemanFirli                       0.0      0.0         0.0            0.0   
franadek                             4.0      4.0         4.0            5.0   
OM INDRA                             3.0      0.0         2.0            0.0   

                  Dua Garis Biru  Avengers: End Game  The Lion King  Aladdin  \
Nama Anda                                                                      
Hania                        4.0                 0.0            0.0      0.0   
Topik Zulkarnain             0.0                 5.0            5.0      0.0   
AhokTemanFirli               0.0                 3.0            0.0      0.0   
franadek                     3.0                 5.0            4.0      5.0   
OM INDRA                     5.0                 5.0            0.0      1.0   

                  Spiderman: Far From Home  Captain Marvel  
Nama Anda                                                   
Hania                                  0.0             0.0  
Topik Zulkarnain                       4.0             2.0  
AhokTemanFirli                         0.0             4.0  
franadek                               4.0             4.0  
OM INDRA                               5.0             5.0

## Calculate Each Person's Similarity Using Eucledian Similarity

### 1. Matrix Similarity Eucledian

In [8]:
from scipy.spatial import distance_matrix
data = data.transpose()
data_euclid = pd.DataFrame(1 / (1 + distance_matrix(data.T, data.T)), columns = col, index = col)
data_euclid

Nama Anda                     Hania  Topik Zulkarnain  AhokTemanFirli  \
Nama Anda                                                               
Hania                      1.000000          0.075025        0.088152   
Topik Zulkarnain           0.075025          1.000000        0.125000   
AhokTemanFirli             0.088152          0.125000        1.000000   
franadek                   0.090094          0.086333        0.077171   
OM INDRA                   0.083020          0.104141        0.107450   
Mulya                      0.075025          0.142857        0.127297   
Genjeh                     0.083651          0.089301        0.093051   
faizah                     0.121786          0.083651        0.093051   
Dpv                        0.086333          0.094414        0.096830   
Putrisqiana                0.114338          0.100560        0.103517   
Nonton_Saat_Diskon         0.074573          0.108152        0.088913   
Romantika                  0.112689          0.083651        0.103517   
ANI                        0.098375          0.081210        0.094882   
Rima                       0.083020          0.078977        0.071039   
bunga                      0.071824          0.108869        0.086333   
Jawaharal                  0.081503          0.097852        0.082403   
Febi ganteng gak ada obat  0.081210          0.092610        0.093051   
Star                       0.075486          0.099449        0.096830   
Indra Junior               0.068834          0.108869        0.086333   
Indra 1991 SM              0.116961          0.093051        0.123899   
Damar Teman Firli          0.081503          0.118818        0.106762   
jul                        0.094882          0.141188        0.128496   
luck                       0.093953          0.122829        0.210897   
Maria O.                   0.082709          0.113504        0.109601   

Nama Anda                  franadek  OM INDRA     Mulya    Genjeh    faizah  \
Nama Anda                                                                     
Hania                      0.090094  0.083020  0.075025  0.083651  0.121786   
Topik Zulkarnain           0.086333  0.104141  0.142857  0.089301  0.083651   
AhokTemanFirli             0.077171  0.107450  0.127297  0.093051  0.093051   
franadek                   1.000000  0.098375  0.096331  0.121786  0.094414   
OM INDRA                   0.098375  1.000000  0.111111  0.111111  0.097337   
Mulya                      0.096331  0.111111  1.000000  0.117878  0.080634   
Genjeh                     0.121786  0.111111  0.117878  1.000000  0.083020   
faizah                     0.094414  0.097337  0.080634  0.083020  1.000000   
Dpv                        0.101711  0.123899  0.094414  0.158945  0.086333   
Putrisqiana                0.126132  0.142857  0.098375  0.119782  0.128496   
Nonton_Saat_Diskon         0.139579  0.101711  0.126132  0.158945  0.076435   
Romantika                  0.075486  0.085638  0.083651  0.090094  0.106762   
ANI                        0.108152  0.093498  0.100560  0.106762  0.098375   
Rima                       0.289898  0.095358  0.087047  0.126132  0.083020   
bunga                      0.120771  0.093051  0.127297  0.120771  0.071824   
Jawaharal                  0.144591  0.122829  0.113504  0.122829  0.080920   
Febi ganteng gak ada obat  0.121786  0.117878  0.126132  0.333333  0.084297   
Star                       0.108152  0.086333  0.112689  0.121786  0.073693   
Indra Junior               0.113504  0.127297  0.127297  0.107450  0.075255   
Indra 1991 SM              0.072227  0.078187  0.093051  0.084626  0.082100   
Damar Teman Firli          0.082100  0.115192  0.110348  0.108869  0.093051   
jul                        0.107450  0.122829  0.166667  0.111891  0.087410   
luck                       0.080920  0.110348  0.118818  0.103517  0.106088   
Maria O.                   0.077674  0.076194  0.122829  0.079792  0.077674   

Nama Anda                       Dpv  Putrisqia

### 2. Show People with High Similarity with "Topik Zulkarnain"

In [9]:
data_nama = pd.DataFrame(data_euclid.loc["Topik Zulkarnain"].sort_values(ascending = False))
data_nama.iloc[1:6]

Topik Zulkarnain
Nama Anda                          
Mulya                      0.142857
jul                        0.141188
AhokTemanFirli             0.125000
luck                       0.122829
Damar Teman Firli          0.118818

### 3. Show People with High Similarity with "Hania"

In [10]:
data_nama = pd.DataFrame(data_euclid.loc["Hania"].sort_values(ascending = False))
data_nama.iloc[1:6]

Hania
Nama Anda              
faizah         0.121786
Indra 1991 SM  0.116961
Putrisqiana    0.114338
Romantika      0.112689
ANI            0.098375

***
Based on the results above, people who have high similarity with "Taufik Zulkarnain" are Mulya (0.142857), jul (0.141188), AhokTemanFirli (0.125000), luck (0.122829) and Damar Teman Firli (0.118818) and people who have high similarity with "Hania" are faizah (0.121786), Indra 1991 SM (0.116961), Putrisqiana (0.114338), Romantika (0.112689) and ANI (0.098375).
***

## Calculate Each Person's Similarity Using Cosine Similarity

### 1. Matrix Similarity Cosine

In [11]:
from sklearn.metrics.pairwise import cosine_similarity
data_cosine = cosine_similarity(data.transpose())
data_cosine = pd.DataFrame(data_cosine, columns = col, index = col)
data_cosine

Nama Anda                     Hania  Topik Zulkarnain  AhokTemanFirli  \
Nama Anda                                                               
Hania                      1.000000          0.000000        0.000000   
Topik Zulkarnain           0.000000          1.000000        0.549805   
AhokTemanFirli             0.000000          0.549805        1.000000   
franadek                   0.658486          0.614701        0.462121   
OM INDRA                   0.382686          0.615689        0.655610   
Mulya                      0.000000          0.742857        0.573710   
Genjeh                     0.451756          0.509753        0.539641   
faizah                     0.717805          0.298807        0.300000   
Dpv                        0.490203          0.572172        0.609272   
Putrisqiana                0.682208          0.550848        0.529514   
Nonton_Saat_Diskon         0.351651          0.741684        0.571548   
Romantika                  0.546608          0.000000        0.000000   
ANI                        0.535453          0.241473        0.323249   
Rima                       0.661224          0.609938        0.476290   
bunga                      0.317282          0.763117        0.558661   
Jawaharal                  0.504787          0.687640        0.488628   
Febi ganteng gak ada obat  0.413309          0.551366        0.539641   
Star                       0.307579          0.624188        0.609272   
Indra Junior               0.246775          0.763117        0.558661   
Indra 1991 SM              0.552158          0.000000        0.000000   
Damar Teman Firli          0.191273          0.621059        0.346410   
jul                        0.375653          0.722806        0.529150   
luck                       0.184560          0.532676        0.742781   
Maria O.                   0.000000          0.466658        0.000000   

Nama Anda                  franadek  OM INDRA     Mulya    Genjeh    faizah  \
Nama Anda                                                                     
Hania                      0.658486  0.382686  0.000000  0.451756  0.717805   
Topik Zulkarnain           0.614701  0.615689  0.742857  0.509753  0.298807   
AhokTemanFirli             0.462121  0.655610  0.573710  0.539641  0.300000   
franadek                   1.000000  0.732994  0.721605  0.843374  0.700635   
OM INDRA                   0.732994  1.000000  0.671660  0.741826  0.599415   
Mulya                      0.721605  0.671660  1.000000  0.759428  0.239046   
Genjeh                     0.843374  0.741826  0.759428  1.000000  0.478714   
faizah                     0.700635  0.599415  0.239046  0.478714  1.000000   
Dpv                        0.759036  0.798889  0.572172  0.893939  0.522233   
Putrisqiana                0.862440  0.835742  0.527408  0.776672  0.774659   
Nonton_Saat_Diskon         0.888528  0.711189  0.839274  0.902550  0.424578   
Romantika                  0.421637  0.331133  0.000000  0.492366  0.565685   
ANI                        0.790569  0.558196  0.531240  0.703380  0.575787   
Rima                       0.988944  0.764719  0.715660  0.888336  0.660001   
bunga                      0.844700  0.657780  0.858507  0.819682  0.359139   
Jawaharal                  0.898757  0.826709  0.800676  0.830014  0.520153   
Febi ganteng gak ada obat  0.843374  0.774434  0.801041  0.984848  0.496121   
Star                       0.791474  0.546180  0.728219  0.803030  0.322044   
Indra Junior               0.820905  0.837174  0.858507  0.764110  0.422986   
Indra 1991 SM              0.298142  0.000000  0.000000  0.348155  0.000000   
Damar Teman Firli          0.559431  0.702959  0.552052  0.703526  0.461880   
jul                        0.816982  0.743392  0.813157  0.723747  0.340168   
luck                       0.553637  0.678287  0.488286  0.694996  0.538516   
Maria O.                   0.465620  0.058508  0.559990  0.271864  0.000000   

Nama Anda                       Dpv  Putrisqia

### 2. Show People with High Similarity with "Topik Zulkarnain"

In [12]:
data_nama = pd.DataFrame(data_cosine.loc["Topik Zulkarnain"].sort_values(ascending = False))
data_nama.iloc[1:6]

Topik Zulkarnain
Nama Anda                           
Indra Junior                0.763117
bunga                       0.763117
Mulya                       0.742857
Nonton_Saat_Diskon          0.741684
jul                         0.722806

### 3. Show People with High Similarity with "Hania"

In [13]:
data_nama = pd.DataFrame(data_cosine.loc["Hania"].sort_values(ascending = False))
data_nama.iloc[1:6]

Hania
Nama Anda              
faizah         0.717805
Putrisqiana    0.682208
Rima           0.661224
franadek       0.658486
Indra 1991 SM  0.552158

***
Based on the results above, people who have high similarity with "Taufik Zulkarnain" are Indra Junior (0.763117), bunga (0.763117), Mulya (0.742857), Nonton_Saat_Diskon (0.741684) and jul (0.722806) and people who have high similarity with "Hania" are faizah (0.717805), Putrisqiana (0.682208), Rima (0.661224), franadek (0.658486) and Indra 1991 SM (0.552158).
***

## Calculate Each Person's Correlation

In [14]:
data_corr = data.corr(method = "pearson")
data_corr

Nama Anda                     Hania  Topik Zulkarnain  AhokTemanFirli  \
Nama Anda                                                               
Hania                      1.000000         -0.741027       -0.481843   
Topik Zulkarnain          -0.741027          1.000000        0.394996   
AhokTemanFirli            -0.481843          0.394996        1.000000   
franadek                  -0.325300          0.142374        0.188092   
OM INDRA                  -0.339788          0.295226        0.550113   
Mulya                     -0.905908          0.567809        0.414680   
Genjeh                    -0.309524          0.023157        0.344173   
faizah                     0.451097         -0.213100       -0.012420   
Dpv                       -0.105381          0.209087        0.469181   
Putrisqiana                0.213478          0.065255        0.326222   
Nonton_Saat_Diskon        -0.762947          0.552692        0.425932   
Romantika                  0.365963         -0.379663       -0.246871   
ANI                        0.000000         -0.434446       -0.028074   
Rima                      -0.294245          0.100164        0.297738   
bunga                     -0.667458          0.588579        0.382716   
Jawaharal                 -0.502906          0.421870        0.231720   
Febi ganteng gak ada obat -0.404762          0.115785        0.344173   
Star                      -0.505830          0.311580        0.469181   
Indra Junior              -0.845447          0.588579        0.382716   
Indra 1991 SM              0.487950         -0.253109       -0.164581   
Damar Teman Firli         -0.319438          0.434959        0.138527   
jul                       -0.532397          0.497096        0.323230   
luck                      -0.303770          0.315142        0.668420   
Maria O.                  -0.484228          0.277343       -0.244988   

Nama Anda                  franadek  OM INDRA     Mulya    Genjeh    faizah  \
Nama Anda                                                                     
Hania                     -0.325300 -0.339788 -0.905908 -0.309524  0.451097   
Topik Zulkarnain           0.142374  0.295226  0.567809  0.023157 -0.213100   
AhokTemanFirli             0.188092  0.550113  0.414680  0.344173 -0.012420   
franadek                   1.000000 -0.324967  0.464140  0.325300  0.117393   
OM INDRA                  -0.324967  1.000000  0.316025  0.294483  0.138971   
Mulya                      0.464140  0.316025  1.000000  0.478118 -0.444977   
Genjeh                     0.325300  0.294483  0.478118  1.000000 -0.236289   
faizah                     0.117393  0.138971 -0.444977 -0.236289  1.000000   
Dpv                       -0.043193  0.505311  0.102466  0.716592 -0.030424   
Putrisqiana                0.041667  0.528071 -0.174052  0.213478  0.478745   
Nonton_Saat_Diskon         0.287554  0.115138  0.689570  0.657713 -0.541166   
Romantika                 -0.166667 -0.023212 -0.464140  0.243975  0.396203   
ANI                        0.348263 -0.078529  0.020526  0.194211  0.091989   
Rima                       0.452267  0.027995  0.404281  0.882735 -0.278739   
bunga                      0.258378  0.046568  0.719540  0.422724 -0.554002   
Jawaharal                 -0.257663  0.472487  0.611245  0.314316 -0.436704   
Febi ganteng gak ada obat  0.325300  0.385093  0.578774  0.952381 -0.193327   
Star                       0.316752 -0.116302  0.436596  0.463677 -0.467765   
Indra Junior              -0.045596  0.554587  0.719540  0.244735 -0.393422   
Indra 1991 SM             -0.111111 -0.402340 -0.309426  0.162650 -0.322832   
Damar Teman Firli          0.145479  0.526787  0.292597  0.532397  0.134491   
jul                       -0.145479  0.283654  0.607701  0.106479 -0.614816   
luck                       0.253629  0.500948  0.206900  0.540035  0.280149   
Maria O.                   0.202150 -0.496529  0.392362 -0.188311 -0.480554   

Nama Anda                       Dpv  Putrisqia

### 2. Show People with High Correlation with "Topik Zulkarnain"

In [15]:
data_nama = pd.DataFrame(data_corr.loc["Topik Zulkarnain"].sort_values(ascending = False))
data_nama.iloc[1:6]

Topik Zulkarnain
Nama Anda                           
Indra Junior                0.588579
bunga                       0.588579
Mulya                       0.567809
Nonton_Saat_Diskon          0.552692
jul                         0.497096

### 3. Show People with High Correlation with "Hania"

In [16]:
data_nama = pd.DataFrame(data_corr.loc["Hania"].sort_values(ascending = False))
data_nama.iloc[1:6]

Hania
Nama Anda              
Indra 1991 SM  0.487950
faizah         0.451097
Romantika      0.365963
Putrisqiana    0.213478
ANI            0.000000

***
Based on the results above, people who have high similarity with "Taufik Zulkarnain" are Indra Junior (0.588579), bunga (0.588579), Mulya (0.567809), Nonton_Saat_Diskon (0.552692) and jul (0.497096) and people who have high similarity with "Hania" are Indra 1991 SM (0.487950), faizah (0.451097), Romantika (0.365963), Putrisqiana (0.213478) and ANI (0.000000).
***

***
Because the results that obtained from Cosine Similarity and Correlation provide a better distribution and give more similar results compared to Eucledian Similarity, we choose to use Cosine Similarity and Correlation for Modelling CF User Based.
***

## Modelling CF User Based Using Cosine Similarity

## 1. Case for Person "Topik Zulkarnain"

### 1.1 Make Matrix Rating from People with Similarity > 0.5 with "Topik Zulkarnain"
* Step 1: Take cosine similarity data with the column "Topik Zulkarnain" and sorted from largest to smallest
* Step 2: Filter data by only taking similarity values that are more than 0.5
* Step 3: Join similarity data with user rating data using the inner method

In [17]:
data_re = pd.DataFrame(data_cosine.loc["Topik Zulkarnain"].sort_values(ascending = False))
data_re = data_re[data_re["Topik Zulkarnain"] > 0.5]
data_re = pd.merge(data_re, data_ori, how = "inner", on = "Nama Anda")
data_re

Topik Zulkarnain  Ada Apa dengan Cinta 2  Gundala  \
Nama Anda                                                                      
Topik Zulkarnain                   1.000000                     0.0      0.0   
Indra Junior                       0.763117                     4.0      0.0   
bunga                              0.763117                     0.0      4.0   
Mulya                              0.742857                     0.0      0.0   
Nonton_Saat_Diskon                 0.741684                     4.0      3.0   
jul                                0.722806                     0.0      3.0   
Jawaharal                          0.687640                     2.0      4.0   
Star                               0.624188                     4.0      4.0   
Damar Teman Firli                  0.621059                     5.0      0.0   
OM INDRA                           0.615689                     3.0      0.0   
franadek                           0.614701                     4.0      4.0   
Rima                               0.609938                     5.0      5.0   
Dpv                                0.572172                     5.0      4.0   
Febi ganteng gak ada obat          0.551366                     4.0      3.0   
Putrisqiana                        0.550848                     4.0      3.0   
AhokTemanFirli                     0.549805                     0.0      0.0   
luck                               0.532676                     3.0      0.0   
Genjeh                             0.509753                     5.0      4.0   

                           Dilan 1991  Bumi Manusia   Dua Garis Biru  \
Nama Anda                                                              
Topik Zulkarnain                  0.0            0.0             0.0   
Indra Junior                      0.0            0.0             4.0   
bunga                             4.0            0.0             0.0   
Mulya                             0.0            0.0             0.0   
Nonton_Saat_Diskon                3.0            0.0             0.0   
jul                               0.0            0.0             3.0   
Jawaharal                         3.0            0.0             5.0   
Star                              0.0            0.0             0.0   
Damar Teman Firli                 0.0            0.0             0.0   
OM INDRA                          2.0            0.0             5.0   
franadek                          4.0            5.0             3.0   
Rima                              5.0            4.0             3.0   
Dpv                               4.0            0.0             0.0   
Febi ganteng gak ada obat         4.0            0.0             0.0   
Putrisqiana                       2.0            4.0             3.0   
AhokTemanFirli                    0.0            0.0             0.0   
luck                              0.0            0.0             0.0   
Genjeh                            3.0            0.0             0.0   

                           Avengers: End Game  The Lion King  Aladdin  \
Nama Anda                                                               
Topik Zulkarnain                          5.0            5.0      0.0   
Indra Junior                              5.0            5.0      5.0   
bunga                                     5.0            5.0      5.0   
Mulya                                     4.0            2.0      5.0   
Nonton_Saat_Diskon                        5.0            5.0      5.0   
jul                                       3.0            3.0      3.0   
Jawaharal                                 5.0            4.0      5.0   
Star                                      5.0            5.0      5.0   
Damar Teman Firli                         5.0            0.0      0.0   
OM INDRA                                  5.0            0.0      1.0   
franadek                                  5.0            4.0      5.0   
Rima                                

### 1.2 Calculate Total Similarity Value
* Step 1: Transpose the data
* Step 2: Take data with the column "Topik Zulkarnain"
* Step 3: Do .sum() method to that column

In [18]:
data_re = data_re.transpose()
data_cosine_film = data_re.iloc[0,1:]
data_cosine_film = pd.DataFrame(data_cosine_film)
sum_cosine = data_cosine_film["Topik Zulkarnain"].sum()
sum_cosine

10.77341631472542

### 1.3 Show the Movie that Hasn't Been Watched by "Topik Zulkarnain"
* Step 1: Dropping the unused column
* Step 2: Take the movies title that was rated "0" by "Topik Zulkarnain"
* Step 3: Dropping the unused row

In [19]:
data_re.drop("Topik Zulkarnain", axis = 0, inplace = True)
data_re = data_re[data_re["Topik Zulkarnain"] == 0]
data_re.drop("Topik Zulkarnain", axis = 1, inplace = True)
data_re

Nama Anda               Indra Junior  bunga  Mulya  Nonton_Saat_Diskon  jul  \
Ada Apa dengan Cinta 2           4.0    0.0    0.0                 4.0  0.0   
Gundala                          0.0    4.0    0.0                 3.0  3.0   
Dilan 1991                       0.0    4.0    0.0                 3.0  0.0   
Bumi Manusia                     0.0    0.0    0.0                 0.0  0.0   
Dua Garis Biru                   4.0    0.0    0.0                 0.0  3.0   
Aladdin                          5.0    5.0    5.0                 5.0  3.0   

Nama Anda               Jawaharal  Star  Damar Teman Firli  OM INDRA  \
Ada Apa dengan Cinta 2        2.0   4.0                5.0       3.0   
Gundala                       4.0   4.0                0.0       0.0   
Dilan 1991                    3.0   0.0                0.0       2.0   
Bumi Manusia                  0.0   0.0                0.0       0.0   
Dua Garis Biru                5.0   0.0                0.0       5.0   
Aladdin                       5.0   5.0                0.0       1.0   

Nama Anda               franadek  Rima  Dpv  Febi ganteng gak ada obat  \
Ada Apa dengan Cinta 2       4.0   5.0  5.0                        4.0   
Gundala                      4.0   5.0  4.0                        3.0   
Dilan 1991                   4.0   5.0  4.0                        4.0   
Bumi Manusia                 5.0   4.0  0.0                        0.0   
Dua Garis Biru               3.0   3.0  0.0                        0.0   
Aladdin                      5.0   5.0  0.0                        5.0   

Nama Anda               Putrisqiana  AhokTemanFirli  luck  Genjeh  
Ada Apa dengan Cinta 2          4.0             0.0   3.0     5.0  
Gundala                         3.0             0.0   0.0     4.0  
Dilan 1991                      2.0             0.0   0.0     3.0  
Bumi Manusia                    4.0             0.0   0.0     0.0  
Dua Garis Biru                  3.0             0.0   0.0     0.0  
Aladdin                         0.0             0.0   0.0     5.0

### 1.4 Show Recommended Movie with Predicted Rating
* Step 1: Do matrix multiplication to similarity data and people rating for movies that "Topik Zulkarnain" hasn't watched data
* Step 2: Divide the matrix multiplication results by total similarity value
* Step 3: Show the movies recommendation with the predicted rating value

In [20]:
data_reccom = data_re.dot(data_cosine_film)
data_reccom["Topik Zulkarnain"] = data_reccom["Topik Zulkarnain"] / sum_cosine
data_reccom.sort_values("Topik Zulkarnain", ascending = False)

Topik Zulkarnain
Aladdin                         3.325399
Ada Apa dengan Cinta 2          2.948795
Gundala                         2.398133
Dilan 1991                      1.968309
Dua Garis Biru                  1.583898
Bumi Manusia                    0.716268

## 2. Case for Person "Hania"

### 2.1 Make Matrix Rating from People with Similarity > 0.5 with "Hania"
* Step 1: Take cosine similarity data with the column "Hania" and sorted from largest to smallest
* Step 2: Filter data by only taking similarity values that are more than 0.5
* Step 3: Join similarity data with user rating data using the inner method

In [21]:
data_re1 = pd.DataFrame(data_cosine.loc["Hania"].sort_values(ascending = False))
data_re1 = data_re1[data_re1["Hania"] > 0.5]
data_re1 = pd.merge(data_re1, data_ori, how = "inner", on = "Nama Anda")
data_re1

Hania  Ada Apa dengan Cinta 2  Gundala  Dilan 1991  \
Nama Anda                                                              
Hania          1.000000                     3.0      5.0         4.0   
faizah         0.717805                     3.0      0.0         5.0   
Putrisqiana    0.682208                     4.0      3.0         2.0   
Rima           0.661224                     5.0      5.0         5.0   
franadek       0.658486                     4.0      4.0         4.0   
Indra 1991 SM  0.552158                     0.0      5.0         0.0   
Romantika      0.546608                     5.0      0.0         5.0   
ANI            0.535453                     4.0      0.0         4.0   
Jawaharal      0.504787                     2.0      4.0         3.0   

               Bumi Manusia   Dua Garis Biru  Avengers: End Game  \
Nama Anda                                                          
Hania                    4.0             4.0                 0.0   
faizah                   5.0             4.0                 5.0   
Putrisqiana              4.0             3.0                 5.0   
Rima                     4.0             3.0                 5.0   
franadek                 5.0             3.0                 5.0   
Indra 1991 SM            0.0             0.0                 0.0   
Romantika                0.0             0.0                 0.0   
ANI                      5.0             0.0                 0.0   
Jawaharal                0.0             5.0                 5.0   

               The Lion King  Aladdin  Spiderman: Far From Home  \
Nama Anda                                                         
Hania                    0.0      0.0                       0.0   
faizah                   0.0      0.0                       0.0   
Putrisqiana              0.0      0.0                       4.0   
Rima                     4.0      5.0                       5.0   
franadek                 4.0      5.0                       4.0   
Indra 1991 SM            0.0      0.0                       0.0   
Romantika                0.0      0.0                       0.0   
ANI                      0.0      4.0                       3.0   
Jawaharal                4.0      5.0                       5.0   

               Captain Marvel  
Nama Anda                      
Hania                     0.0  
faizah                    0.0  
Putrisqiana               3.0  
Rima                      5.0  
franadek                  4.0  
Indra 1991 SM             0.0  
Romantika                 0.0  
ANI                       4.0  
Jawaharal                 4.0

### 2.2 Calculate Total Similarity Value
* Step 1: Transpose the data
* Step 2: Take data with the column "Hania"
* Step 3: Do .sum() method to that column

In [22]:
data_re1 = data_re1.transpose()
data_cosine_film = data_re1.iloc[0,1:]
data_cosine_film = pd.DataFrame(data_cosine_film)
sum_cosine = data_cosine_film["Hania"].sum()
sum_cosine

4.858729602053706

### 2.3 Show the Movie that Hasn't Been Watched by "Hania"
* Step 1: Dropping the unused column
* Step 2: Take the movies title that was rated "0" by "Hania"
* Step 3: Dropping the unused row

In [23]:
data_re1.drop("Hania", axis = 0, inplace = True)
data_re1 = data_re1[data_re1["Hania"] == 0]
data_re1.drop("Hania", axis = 1, inplace = True)
data_re1

Nama Anda                 faizah  Putrisqiana  Rima  franadek  Indra 1991 SM  \
Avengers: End Game           5.0          5.0   5.0       5.0            0.0   
The Lion King                0.0          0.0   4.0       4.0            0.0   
Aladdin                      0.0          0.0   5.0       5.0            0.0   
Spiderman: Far From Home     0.0          4.0   5.0       4.0            0.0   
Captain Marvel               0.0          3.0   5.0       4.0            0.0   

Nama Anda                 Romantika  ANI  Jawaharal  
Avengers: End Game              0.0  0.0        5.0  
The Lion King                   0.0  0.0        4.0  
Aladdin                         0.0  4.0        5.0  
Spiderman: Far From Home        0.0  3.0        5.0  
Captain Marvel                  0.0  4.0        4.0

### 2.4 Show Recommended Movie with Predicted Rating
* Step 1: Do matrix multiplication to similarity data and people rating for movies that "Hania" hasn't watched data
* Step 2: Divide the matrix multiplication results by total similarity value
* Step 3: Show the movies recommendation with the predicted rating value

In [24]:
data_reccom1 = data_re1.dot(data_cosine_film)
data_reccom1["Hania"] = data_reccom1["Hania"] / sum_cosine
data_reccom1.sort_values("Hania", ascending = False)

Hania
Avengers: End Game        3.318265
Spiderman: Far From Home  2.634268
Captain Marvel            2.500170
Aladdin                   2.318363
The Lion King             1.502037

***
Based on the results above, there are 6 recommended movies for "Topik Zulkarnain" including Aladdin (3.325399), Ada Apa dengan Cinta 2 (2.948795), Gundala (2.398133), Dilan 1991 (1.968309), Dua Garis Biru (1.583898) and Bumi Manusia (0.716268) also there are 5 recommended movies for "Hania" including Avengers: End Game (3.318265), Spiderman: Far From Home (2.634268), Captain Marvel (2.500170), Aladdin (2.318363) and The Lion King (1.502037).
***

## Modelling CF User Based Using Correlation

## 1. Case for Person "Topik Zulkarnain"

### 1.1 Make Matrix Rating from People with Correlation > 0 with "Topik Zulkarnain"
* Step 1: Take correlation data with the column "Topik Zulkarnain" and sorted from largest to smallest
* Step 2: Filter data by only taking correlation values that are more than 0
* Step 3: Join correlation data with user rating data using the inner method

In [25]:
data_rec = pd.DataFrame(data_corr.loc["Topik Zulkarnain"].sort_values(ascending = False))
data_rec = data_rec[data_rec["Topik Zulkarnain"] > 0]
data_rec = pd.merge(data_rec, data_ori, how = "inner", on = "Nama Anda")
data_rec

Topik Zulkarnain  Ada Apa dengan Cinta 2  Gundala  \
Nama Anda                                                                      
Topik Zulkarnain                   1.000000                     0.0      0.0   
Indra Junior                       0.588579                     4.0      0.0   
bunga                              0.588579                     0.0      4.0   
Mulya                              0.567809                     0.0      0.0   
Nonton_Saat_Diskon                 0.552692                     4.0      3.0   
jul                                0.497096                     0.0      3.0   
Damar Teman Firli                  0.434959                     5.0      0.0   
Jawaharal                          0.421870                     2.0      4.0   
AhokTemanFirli                     0.394996                     0.0      0.0   
luck                               0.315142                     3.0      0.0   
Star                               0.311580                     4.0      4.0   
OM INDRA                           0.295226                     3.0      0.0   
Maria O.                           0.277343                     0.0      0.0   
Dpv                                0.209087                     5.0      4.0   
franadek                           0.142374                     4.0      4.0   
Febi ganteng gak ada obat          0.115785                     4.0      3.0   
Rima                               0.100164                     5.0      5.0   
Putrisqiana                        0.065255                     4.0      3.0   
Genjeh                             0.023157                     5.0      4.0   

                           Dilan 1991  Bumi Manusia   Dua Garis Biru  \
Nama Anda                                                              
Topik Zulkarnain                  0.0            0.0             0.0   
Indra Junior                      0.0            0.0             4.0   
bunga                             4.0            0.0             0.0   
Mulya                             0.0            0.0             0.0   
Nonton_Saat_Diskon                3.0            0.0             0.0   
jul                               0.0            0.0             3.0   
Damar Teman Firli                 0.0            0.0             0.0   
Jawaharal                         3.0            0.0             5.0   
AhokTemanFirli                    0.0            0.0             0.0   
luck                              0.0            0.0             0.0   
Star                              0.0            0.0             0.0   
OM INDRA                          2.0            0.0             5.0   
Maria O.                          0.0            0.0             0.0   
Dpv                               4.0            0.0             0.0   
franadek                          4.0            5.0             3.0   
Febi ganteng gak ada obat         4.0            0.0             0.0   
Rima                              5.0            4.0             3.0   
Putrisqiana                       2.0            4.0             3.0   
Genjeh                            3.0            0.0             0.0   

                           Avengers: End Game  The Lion King  Aladdin  \
Nama Anda                                                               
Topik Zulkarnain                          5.0            5.0      0.0   
Indra Junior                              5.0            5.0      5.0   
bunga                                     5.0            5.0      5.0   
Mulya                                     4.0            2.0      5.0   
Nonton_Saat_Diskon                        5.0            5.0      5.0   
jul                                       3.0            3.0      3.0   
Damar Teman Firli                         5.0            0.0      0.0   
Jawaharal                                 5.0            4.0      5.0   
AhokTemanFirli                            3.0            0.0      0.0   
luck                          

### 1.2 Calculate Total Correlation Value
* Step 1: Transpose the data
* Step 2: Take data with the column "Topik Zulkarnain"
* Step 3: Do .sum() method to that column

In [26]:
data_rec = data_rec.transpose()
data_corr_film = data_rec.iloc[0,1:]
data_corr_film = pd.DataFrame(data_corr_film)
sum_cor = data_corr_film["Topik Zulkarnain"].sum()
sum_cor

5.901692203591875

### 1.3 Show the Movie that Hasn't Been Watched by "Topik Zulkarnain"
* Step 1: Dropping the unused column
* Step 2: Take the movies title that was rated "0" by "Topik Zulkarnain"
* Step 3: Dropping the unused row

In [27]:
data_rec.drop("Topik Zulkarnain", axis = 0, inplace = True)
data_rec = data_rec[data_rec["Topik Zulkarnain"] == 0]
data_rec.drop("Topik Zulkarnain", axis = 1, inplace = True)
data_rec

Nama Anda               Indra Junior  bunga  Mulya  Nonton_Saat_Diskon  jul  \
Ada Apa dengan Cinta 2           4.0    0.0    0.0                 4.0  0.0   
Gundala                          0.0    4.0    0.0                 3.0  3.0   
Dilan 1991                       0.0    4.0    0.0                 3.0  0.0   
Bumi Manusia                     0.0    0.0    0.0                 0.0  0.0   
Dua Garis Biru                   4.0    0.0    0.0                 0.0  3.0   
Aladdin                          5.0    5.0    5.0                 5.0  3.0   

Nama Anda               Damar Teman Firli  Jawaharal  AhokTemanFirli  luck  \
Ada Apa dengan Cinta 2                5.0        2.0             0.0   3.0   
Gundala                               0.0        4.0             0.0   0.0   
Dilan 1991                            0.0        3.0             0.0   0.0   
Bumi Manusia                          0.0        0.0             0.0   0.0   
Dua Garis Biru                        0.0        5.0             0.0   0.0   
Aladdin                               0.0        5.0             0.0   0.0   

Nama Anda               Star  OM INDRA  Maria O.  Dpv  franadek  \
Ada Apa dengan Cinta 2   4.0       3.0       0.0  5.0       4.0   
Gundala                  4.0       0.0       0.0  4.0       4.0   
Dilan 1991               0.0       2.0       0.0  4.0       4.0   
Bumi Manusia             0.0       0.0       0.0  0.0       5.0   
Dua Garis Biru           0.0       5.0       0.0  0.0       3.0   
Aladdin                  5.0       1.0       4.0  0.0       5.0   

Nama Anda               Febi ganteng gak ada obat  Rima  Putrisqiana  Genjeh  
Ada Apa dengan Cinta 2                        4.0   5.0          4.0     5.0  
Gundala                                       3.0   5.0          3.0     4.0  
Dilan 1991                                    4.0   5.0          2.0     3.0  
Bumi Manusia                                  0.0   4.0          4.0     0.0  
Dua Garis Biru                                0.0   3.0          3.0     0.0  
Aladdin                                       5.0   5.0          0.0     5.0

### 1.4 Show Recommended Movie with Predicted Rating

* Step 1: Do matrix multiplication to correlation data and people rating for movies that "Topik Zulkarnain" hasn't watched data
* Step 2: Divide the matrix multiplication results by total correlation value
* Step 3: Show the movies recommendation with the predicted rating value

In [28]:
data_reccomend = data_rec.dot(data_corr_film)
data_reccomend["Topik Zulkarnain"] = data_reccomend["Topik Zulkarnain"] / sum_cor
data_reccomend.sort_values("Topik Zulkarnain", ascending = False)

Topik Zulkarnain
Aladdin                         3.381883
Ada Apa dengan Cinta 2          2.307259
Gundala                         1.860465
Dilan 1991                      1.429800
Dua Garis Biru                  1.415604
Bumi Manusia                    0.232737

## 2. Case for Person "Hania"

### 2.1 Make Matrix Rating from People with Correlation > 0 with "Hania"
* Step 1: Take correlation data with the column "Hania" and sorted from largest to smallest
* Step 2: Filter data by only taking correlation values that are more than 0
* Step 3: Join correlation data with user rating data using the inner method

In [29]:
data_rec1 = pd.DataFrame(data_corr.loc["Hania"].sort_values(ascending = False))
data_rec1 = data_rec1[data_rec1["Hania"] > 0]
data_rec1 = pd.merge(data_rec1, data_ori, how = "inner", on = "Nama Anda")
data_rec1

Hania  Ada Apa dengan Cinta 2  Gundala  Dilan 1991  \
Nama Anda                                                              
Hania          1.000000                     3.0      5.0         4.0   
Indra 1991 SM  0.487950                     0.0      5.0         0.0   
faizah         0.451097                     3.0      0.0         5.0   
Romantika      0.365963                     5.0      0.0         5.0   
Putrisqiana    0.213478                     4.0      3.0         2.0   

               Bumi Manusia   Dua Garis Biru  Avengers: End Game  \
Nama Anda                                                          
Hania                    4.0             4.0                 0.0   
Indra 1991 SM            0.0             0.0                 0.0   
faizah                   5.0             4.0                 5.0   
Romantika                0.0             0.0                 0.0   
Putrisqiana              4.0             3.0                 5.0   

               The Lion King  Aladdin  Spiderman: Far From Home  \
Nama Anda                                                         
Hania                    0.0      0.0                       0.0   
Indra 1991 SM            0.0      0.0                       0.0   
faizah                   0.0      0.0                       0.0   
Romantika                0.0      0.0                       0.0   
Putrisqiana              0.0      0.0                       4.0   

               Captain Marvel  
Nama Anda                      
Hania                     0.0  
Indra 1991 SM             0.0  
faizah                    0.0  
Romantika                 0.0  
Putrisqiana               3.0

### 2.2 Calculate Total Correlation Value
* Step 1: Transpose the data
* Step 2: Take data with the column "Hania"
* Step 3: Do .sum() method to that column

In [30]:
data_rec1 = data_rec1.transpose()
data_corr_film = data_rec1.iloc[0,1:]
data_corr_film = pd.DataFrame(data_corr_film)
sum_cor1 = data_corr_film["Hania"].sum()
sum_cor1

1.5184875592356168

### 2.3 Show the Movie that Hasn't Been Watched by "Hania"
* Step 1: Dropping the unused column
* Step 2: Take the movies title that was rated "0" by "Hania"
* Step 3: Dropping the unused row

In [31]:
data_rec1.drop("Hania", axis = 0, inplace = True)
data_rec1 = data_rec1[data_rec1["Hania"] == 0]
data_rec1.drop("Hania", axis = 1, inplace = True)
data_rec1

Nama Anda                 Indra 1991 SM  faizah  Romantika  Putrisqiana
Avengers: End Game                  0.0     5.0        0.0          5.0
The Lion King                       0.0     0.0        0.0          0.0
Aladdin                             0.0     0.0        0.0          0.0
Spiderman: Far From Home            0.0     0.0        0.0          4.0
Captain Marvel                      0.0     0.0        0.0          3.0

### 2.4 Show Recommended Movie with Predicted Rating
* Step 1: Do matrix multiplication to correlation data and people rating for movies that "Hania" hasn't watched data
* Step 2: Divide the matrix multiplication results by total correlation value
* Step 3: Show the movies recommendation with the predicted rating value

In [32]:
data_reccomend1 = data_rec1.dot(data_corr_film)
data_reccomend1["Hania"] = data_reccomend1["Hania"] / sum_cor1
data_reccomend1.sort_values("Hania", ascending = False)

Hania
Avengers: End Game        2.188279
Spiderman: Far From Home  0.562344
Captain Marvel            0.421758
The Lion King             0.000000
Aladdin                   0.000000

***
Based on the results above, there are 6 recommended movies for "Topik Zulkarnain" including Aladdin (3.381883), Ada Apa dengan Cinta 2 (2.307259), Gundala (1.860465), Dilan 1991 (1.429800), Dua Garis Biru (1.415604) and Bumi Manusia (0.232737) also there are 5 recommended movies for "Hania" including Avengers: End Game (2.188279), Spiderman: Far From Home (0.562344), Captain Marvel (0.421758), Aladdin (0.000000) and The Lion King (0.000000).
***

## Conclusion

The results that obtained from the two modelling are not much different, the recommended movies have the same order based on predict rating but the values are different. To find out which model is more optimal we can separate the dataset into training data and testing data where both models are used to predict the rating results on testing data, so that we can obtain the difference in error value. In the real case, we can apply this algorithm to display the movies recommendation on the main page of a movie website when someone has logged in on the website.

***
# THANK YOU